---
layout: page
title: RNA-seq gene-level analysis
---

```{r options, echo=FALSE}
library(knitr)
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

## Introduction

RNA-Seq is a valuable experiment for quantifying both the types and the amount of RNA molecules in a sample. We've covered the basic idea of the protocol in lectures, but some early references for RNA-Seq include [Mortazavi (2008)](#foot) and [Marioni (2008)](#foot).

In this lab, we will focus on comparing the expression levels of genes across different samples, by counting the number of reads which overlap the exons of genes defined by a known annotation. As described in the lecture, this analysis sets aside the task of estimating the different kinds of RNA molecules, and the different isoforms for genes with multiple isoforms. One advantage of looking at these matrices of raw counts is that we can use statistical distributions to model how the variance of counts will change when the counts are low vs high. We will explore the relationship of the variance of counts to the mean later in this lab.

## Counting reads in genes

In this lab we will examine 8 samples from the airway package, which are from the paper by [Himes et al](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4057123/): "RNA-Seq Transcriptome Profiling Identifies CRISPLD2 as a Glucocorticoid Responsive Gene that Modulates Cytokine Function in Airway Smooth Muscle Cells".

This lab will focus on a summarized version of an RNA-seq experiment: a count matrix, which has genes along the rows and samples along the columns. The values in the matrix are the number of reads which could be uniquely aligned to the exons of a given gene for a given sample. We will demonstrate how to build a count matrix for a subset of reads from an experiment, and then use a pre-made count matrix, to avoid having students download the multi-gigabyte BAM files containing the aligned reads. 

```{r}
library(airway)
dir <- system.file("extdata", package="airway", mustWork=TRUE)
csv.file <- file.path(dir, "sample_table.csv")
sample.table <- read.csv(csv.file, row.names=1)
bam.files <- file.path(dir, paste0(sample.table$Run, "_subset.bam"))
gtf.file <- file.path(dir, "Homo_sapiens.GRCh37.75_subset.gtf")
```

```{r}
library(Rsamtools)
bam.list <- BamFileList(bam.files)
library(GenomicFeatures)
# this will be renamed makeTxDb... in Bioc 3.1
txdb <- makeTranscriptDbFromGFF(gtf.file, format="gtf")
exons.by.gene <- exonsBy(txdb, by="gene")
```

```{r cache=TRUE}
library(GenomicAlignments)
se <- summarizeOverlaps(exons.by.gene, bam.list,
                        mode="Union",
                        singleEnd=FALSE,
                        ignore.strand=TRUE,
                        fragments=TRUE)
colData(se) <- DataFrame(sample.table)
```

```{r cache=TRUE}
library(Rsubread)
fc <- featureCounts(bam.files, annot.ext=gtf.file,
                    isGTFAnnotationFile=TRUE, 
                    isPaired=TRUE)
names(fc)
fc$counts
```

```{r}
plot(assay(se)[,1], 
     fc$counts[match(rownames(se),rownames(fc$counts)),1])
abline(0,1)
```

## Visualizing sample-sample distances

```{r, message=FALSE}
library(airway)
data(airway)
airway
```

```{r}
colData(airway)
```

```{r}
# rowRanges for Bioc >= 3.1
rowData(airway)
```

```{r}
head(assay(airway))
```

### Creating a DESeqDataSet object

We will use the `DESeq2` package to normalize the sample for sequencing depth. The *DESeqDataSet* object is just an extension of the *SummarizedExperiment* object. We specify an experimental *design* here, for later use, although for estimating size factors, we could just use `~ 1` as a default design. The variables are columns of the `colData`, and the `+` indicates that for differential expression analysis we want to compare levels of `dex` while controlling for the `cell` differences.

```{r}
library(DESeq2)
dds <- DESeqDataSet(airway, design= ~ cell + dex)
```
We can also make a *DESeqDataSet* from a count matrix and column data.

```{r}
dds2 <- DESeqDataSetFromMatrix(fc$counts, 
                               colData=sample.table, 
                               design=~ cell + dex)
```



This code chunk is not necessary, but helps to make nicer plots below with large axis labels (`mypar(1,2)` can be substituted with `par(mfrow=c(1,2))` below).

```{r}
# library(devtools)
# install_github("ririzarr/rafalib")
library(rafalib)
mypar()
```

### Normalization for sequencing depth

The following estimates size factors to account for differences in sequencing depth, and is only necessary to make the `log.norm.counts` object below.

```{r}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
colSums(counts(dds))
plot(sizeFactors(dds), colSums(counts(dds)))
abline(lm(colSums(counts(dds)) ~ sizeFactors(dds) + 0))
```

Size factors are calculated by the median ratio of samples to a pseudo-sample (the geometric mean of all samples). In other words, for each sample, we take the exponent of the median of the log ratios in this histogram.

```{r}
loggeomeans <- rowMeans(log(counts(dds)))
hist(log(counts(dds)[,1]) - loggeomeans, 
     col="grey", main="", xlab="", breaks=40)
```

Make a matrix of log normalized counts:

```{r}
log.norm.counts <- log2(counts(dds, normalized=TRUE) + 1)
```

```{r}
rs <- rowSums(counts(dds))
mypar(1,2)
boxplot(log2(counts(dds)[rs > 0,]+1)) # not normalized
boxplot(log.norm.counts[rs > 0,]) # normalized
```

```{r}
plot(log.norm.counts[,1:2], cex=.1)
```


### Stabilizing count variance

Now we will use a more sophisticated transformation, which is similar to the variance stablizing normalization method taught in Week 3 of Course 4: Introduction to Bioconductor. It uses the variance model for count data to shrink together the values for genes with very low counts; for genes with medium and high counts, the `rlog` is very close to `log2`.

```{r}
rld <- rlog(dds)
```

```{r}
plot(assay(rld)[,1], assay(rld)[,2], cex=.1)
```

```{r}
library(vsn)
mypar(1,2)
meanSdPlot(log.norm.counts, ranks=FALSE, ylim=c(0,3), main="log2")
meanSdPlot(assay(rld), ranks=FALSE, ylim=c(0,3), main="rlog")
```

Principal components plots:

```{r}
mypar()
rv <- apply(log.norm.counts, 1, var)
topgenes <- head(order(rv, decreasing=TRUE),500)
pc <- prcomp(t(log.norm.counts[topgenes,]))
plot(pc$x[,1], pc$x[,2], 
     col = colData(dds)$dex, 
     pch = as.integer(colData(dds)$cell))
```

```{r}
plotPCA(rld, intgroup="dex")
plotPCA(rld, intgroup=c("dex","cell"))
```

We can make this plot even nicer using custom code from the *ggplot2* library:

```{r}
library(ggplot2)
(data <- plotPCA(rld, intgroup=c("dex","cell"), returnData=TRUE))
(percentVar <- 100*round(attr(data, "percentVar"),2))
```

```{r}
makeLab <- function(x,pc) paste0("PC",pc,": ",x,"% variance")
ggplot(data, aes(PC1,PC2,col=dex,shape=cell)) + geom_point() +
  xlab(makeLab(percentVar[1],1)) + ylab(makeLab(percentVar[2],2))
```

```{r}
mypar(1,2)
plot(hclust(dist(t(log.norm.counts))), labels=colData(dds)$dex)
plot(hclust(dist(t(assay(rld)))), labels=colData(rld)$dex)
```

## Differential gene expression

A number of methods for assessing differential gene expression from RNA-Seq counts use the Negative Binomial distribution to make probabilistic statements about the differences seen in an experiment. A few such methods are [edgeR](#foot), [DESeq](#foot), [DSS](#foot) and *many* others. A very incomplete list of other methods is provided in the [footnotes](#foot).

We will use `DESeq2` to perform differential gene expression on the counts. This also uses a Negative Binomial distribution to model the counts. It performs a similar step to `limma`, in using the variance of all the genes to improve the variance estimate for each individual gene. In addition, it shrinks the unreliable fold changes from genes with low counts, which will be seen in the resulting MA-plot.

First, we setup the `design` of the experiment, so that differences will be considered across time and protocol variables. The last variable is used for the default results tables and plots, and we make sure the "control" level is the first level, such that log fold changes will be treatment over control, and not control over treatment.

```{r}
levels(dds$dex)
dds$dex <- relevel(dds$dex, "untrt")
levels(dds$dex)
design(dds)
design(dds) <- ~ cell + dex
```

The following line runs the model, and then we can extract a results table for all genes:

```{r}
dds <- DESeq(dds)
res <- results(dds)
head(res)
table(res$padj < 0.1)
```

```{r}
summary(res)
```

```{r}
res2 <- results(dds, alpha=0.01)
table(res$padj < 0.01)
```

```{r}
results(dds, contrast=c("cell","N61311","N052611"))
```

```{r}
plotMA(res)
```

Sorted results table:

```{r}
resSort <- res[order(res$pvalue),]
head(resSort)
```

Let's examine the top gene, sorting by p-value:

```{r}
plotCounts(dds, gene=which.min(res$pvalue), intgroup="dex")
```

```{r}
data <- plotCounts(dds, gene=which.min(res$pvalue), intgroup=c("dex","cell"), returnData=TRUE)
ggplot(data, aes(x=dex, y=count, col=cell)) +
  geom_point(position=position_jitter(width=.1,height=0)) +
  scale_y_log10()
```


```{r}
ggplot(data, aes(x=dex, y=count, col=cell, group=cell)) +
  geom_point() + geom_line() + scale_y_log10() 
```

```{r}
library(pheatmap)
topgenes <- head(rownames(resSort),20)
mat <- assay(rld)[topgenes,]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds)[,c("dex","cell")])
pheatmap(mat, annotation_col=df)
```

We can then check the annotation of these highly significant genes:

```{r}
library(org.Hs.eg.db)
keytypes(org.Hs.eg.db)
head(rownames(dds))
anno <- select(org.Hs.eg.db, keys=topgenes,
               columns=c("ENSEMBL","SYMBOL","GENENAME"), 
               keytype="ENSEMBL")
anno
sum(duplicated(anno[,1])) # remember, 1-to-many mappings
# for Bioconductor >= 3.1, use mapIds() function
```

## Footnotes <a name="foot"></a>

### RNA-seq introductory papers

Mortazavi A, Williams BA, McCue K, Schaeffer L, Wold B., "Mapping and quantifying mammalian transcriptomes by RNA-Seq", Nat Methods. 2008.
<http://www.nature.com/nmeth/journal/v5/n7/full/nmeth.1226.html>

John C. Marioni, Christopher E. Mason, Shrikant M. Mane, Matthew Stephens, and Yoav Gilad, "RNA-seq: An assessment of technical reproducibility and comparison with gene expression arrays" Genome Res. 2008.
<http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2527709/>

Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L.,  "Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation", Nature Biotechnology, 2010.
http://www.nature.com/nbt/journal/v28/n5/full/nbt.1621.html

### ReCount

Frazee AC, Langmead B, Leek JT. "ReCount: a multi-experiment resource of analysis-ready RNA-seq gene count datasets". BMC Bioinformatics 12:449 http://www.ncbi.nlm.nih.gov/pubmed/22087737

### Negative Binomial methods for differential expression of count data

All the following methods are available on Bioconductor:

- `edgeR`

Mark D. Robinson, Davis J. McCarthy, and Gordon K. Smyth, "edgeR: a Bioconductor package for differential expression analysis of digital gene expression data" Bioinformatics 2010.
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2796818/

- `DESeq` of which the latest version is a separate package, `DESeq2`. The paper for the latest version *DESeq2* is from 2014:

Michael I Love, Simon Anders, Wolfgang Huber, "Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2"
http://genomebiology.com/2014/15/12/550

The original *DESeq* paper from 2010:

Simon Anders and Wolfgang Huber, "Differential expression analysis for sequence count data", Genome Biology 2010.
http://genomebiology.com/2010/11/10/r106

- `DSS`

Hao Wu, Chi Wang, Zhijin Wu, "A new shrinkage estimator for dispersion improves differential expression detection in RNA-seq data" Biostatistics 2013.
http://biostatistics.oxfordjournals.org/content/14/2/232

### Transformation followed by linear model methods

`voom` in the `limma` Bioconductor package. Limma also contains gene-set testing methods (see ROAST for example in the Reference Manual)

Charity W Law, Yunshun Chen, Wei Shi and Gordon K Smyth, "voom: precision weights unlock linear model analysis tools for RNA-seq read counts", Genome Biology. 2014.
http://genomebiology.com/2014/15/2/R29

### Resampling-based methods

`SAMseq` in the `samr` package on CRAN

Jun Li and Robert Tibshirani, "Finding consistent patterns: A nonparametric approach for identifying differential expression in RNA-Seq data", Stat Methods Med Res. 2013.
http://smm.sagepub.com/content/22/5/519.short

### Incorporating isoform-abundance

- `Cuffdiff` (the latest version is `Cuffdiff2`) with `cummeRbund` the accompanying Bioconductor visualization package.

Trapnell C, Hendrickson DG, Sauvageau M, Goff L, Rinn JL, Pachter L., "Differential analysis of gene regulation at transcript resolution with RNA-seq" Nat Biotechnol. 2013.
http://www.ncbi.nlm.nih.gov/pubmed/23222703

- `BitSeq` (Bioconductor)

Peter Glaus, Antti Honkela, and Magnus Rattray, "Identifying differentially expressed transcripts from RNA-seq data with biological variation", Bioinformatics. 2012.
http://bioinformatics.oxfordjournals.org/content/28/13/1721

