---
title: "Estrogen receptor binding: data representation and location analysis"
author: "Vince"
date: "March 19, 2015"
output: html_document
layout: page
toc: yes
---

```{r options, echo=FALSE}
library(knitr)
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

```{r setup,echo=FALSE,results="hide",message=FALSE}
library(GenomicFeatures)
library(GenomicRanges)
library(IRanges)
library(devtools)
library(ERBS)
library(Homo.sapiens)
```

# Interval data on ER binding

A 2006 Nature Genetics paper (v38, n11) by Carroll, Meyer, Song et al.
describes the notion that a group of genes
whose expression correlates 
with the expression of the estrogen receptor (ER) gene
forms a signature of a breast cancer subtype
called "luminal".  This finding motivates interest in determining
the set of all genomic elements to which ER binds, and this has
been carried out using ChIP-seq experiments.  An important 
finding of the Carroll et al. paper was that ER binding 
in the MCF-7 cell line occurred only rarely promoter-proximal regions.
This study therefore unearthed previously unexplored 
cis-regulatory sites.

We will show how to relate binding peak locations to gene
transcription start sites in two cell lines analyzed in the
ENCODE project.

# Acquisition of binding data

Get the ERBS package.

```{r geter}
library(devtools)
install_github("genomicsclass/ERBS")
library(ERBS)
data(HepG2)  # cell line of liver origin
data(GM12878)  # immortalized B cell
HepG2  # locations of ER binding peaks
```

We will use `granges` for clarify avoid all the columns
```{r}
HepG2
granges(HepG2)
```


# Genomic Ranges tools

```{r}

browseVignettes("GenomicRanges")

res = findOverlaps(HepG2,GM12878)

erbs = HepG2[queryHits(res)]

erbs = granges(erbs)
```


# Acquisition of gene transcription start sites

We can use Homo.sapiens:

```{r geths}
library(Homo.sapiens)
ghs = genes(Homo.sapiens)
ghs
```

Explain genes have direction and what is a transcription start site

```{r}
tssgr = resize(ghs,1)
granges(ghs)[1:3,]
granges(tssgr)[1:3,]
```


```{r}
index <- precede(HepG2,ghs)
HepG2[1:3,]
granges(HepG2)[1:3,]
granges(ghs)[index[1:3],]
d = distance(HepG2,ghs[index])
```

If we want the closest to a start site, regardless of befor or after, for example if overlaps. We want something different.

We can create a granges of TSS like this:

```{r}
tssgr = resize(ghs,1)
```


# Finding nearest gene for each binding event

The distanceToNearest function from GenomicFeatures will
create a Hits object with metadata giving the positive distance
between peak location and nearest TSS.  We'll discuss the Hits
object in the video.

```{r donear}
d = distanceToNearest(HepG2, tssgr)
d
##note this gives an error
try(d[,3])
###instead we have to do this
dists = values(d)$distance
```

We will call a distance negative if the peak is 5' to the nearest TSS.
The density estimate given below shows that the vast majority of
events are remote from the 1kb region around TSS; we use dashed
lines to denote that region.

```{r getsi,fig=TRUE}
index = subjectHits(d)
sdists = ifelse(end(HepG2) < start(tssgr[index]), dists, -dists)
hist(sdists, xlim=c(-100000,100000), main="Density of d(ER binding peak, nearest TSS)" ,breaks=seq(min(sdists),max(sdists),len=1000))
abline(v=-c(10000,10000), lty=2)
```

## Getting information about genes

# Annotating genes

Get information on genes

```{r}
index <- subjectHits(d)[dists<1000]
dists = values(d)$distance
##better way to do this?
geneids <- mcols(tssgr[index])$GENEID
```

```{r geths2}
library(Homo.sapiens)
?select
columns(Homo.sapiens)
keytypes(Homo.sapiens)
geneids <- as.character(geneids)
select(Homo.sapiens,keys=geneids,columns="GENENAME",keytype="GENEID")
```






